File: src/argamak_ffi.mjs
import * as tf from "@tensorflow/tfjs-node";
import { Error as GleamError, List, Ok, Result, toList } from "./gleam.mjs";
import {
  CannotBroadcast,
  IncompatibleShape,
  InvalidData,
} from "./argamak/tensor.mjs";

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
// Constants                              //
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//

const Nil = undefined;

const Tensor = tf.Tensor.prototype;

const Extrema = {
  int32: { min: -2_147_483_648, max: 2_147_483_647 },
  float32: {
    min: -340_282_346_638_528_859_811_704_183_484_516_925_440,
    max: 340_282_346_638_528_859_811_704_183_484_516_925_440,
  },
};

class Fn {
  constructor(f) {
    this.f = f;
  }

  //~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
  // Tensor Methods                       //
  //~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//

  static clip_reformat_like(x) {
    return x.dtype === this.dtype
      // Don't clip if reformat is noop.
      ? this
      : this.reformat_like(x).clip_based_on(this);
  }

  static clip_reformat(x) {
    return (x === this.dtype ? this : this.cast(x)).clip_based_on(this);
  }

  static clip_based_on(x) {
    let format = this.dtype;

    let scalar = (x) => tf.scalar(Extrema[format][x], format);

    let min = scalar("min");
    let max = scalar("max");

    let less = x.less(min);
    let greater = x.greater(max);

    Tensor.replace = function (predicate, value) {
      return value.where(predicate, this);
    };

    return this.replace(less, min).replace(greater, max);
  }

  static reformat_like(x) {
    x = x.dtype;
    return x === this.dtype ? this : this.cast(x);
  }

  static to_number() {
    if (Number.isFinite(this)) {
      return this;
    } else if (tf.util.isScalarShape(this.shape)) {
      return this.arraySync();
    }
    throw new Error(Nil);
  }

  static to_flat_list() {
    return toList(this.reshape([-1]).arraySync());
  }

  static add_fn(x) {
    this[x] = Fn[x];
    return this;
  }

  //~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
  // Result Methods                       //
  //~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//

  static map(f) {
    return this.isOk() ? new Ok(f(this[0])) : this;
  }

  //~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
  // Fn Methods                           //
  //~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//

  broadcast_result() {
    return this.result(CannotBroadcast);
  }

  shape_result() {
    return this.result(IncompatibleShape);
  }

  result(error_type = InvalidData) {
    try {
      let x = this.f();
      // Detect any NaN
      return tf.equal(x, x).all().arraySync()
        ? new Ok(x)
        : new GleamError(new InvalidData());
    } catch {
      return new GleamError(new error_type());
    }
  }
}

Tensor.add_fn = Fn.add_fn;

Tensor.add_fn("clip_reformat_like")
  .add_fn("clip_reformat")
  .add_fn("clip_based_on")
  .add_fn("reformat_like")
  .add_fn("to_number")
  .add_fn("to_flat_list");

Result.prototype.map = Fn.map;

const fn = (f) => new Fn(f);

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
// Tensor Creation Functions              //
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//

export const tensor = (x, format) =>
  fn(() => {
    if (List.isList(x)) {
      x = x.toArray();
      if (!x.length) {
        throw new Error(Nil);
      }
    }
    let shape = Array.isArray(x) ? [x.length] : [];
    return tf.tensor(x, shape, format);
  }).result();

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
// Tensor Reflection Functions            //
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//

export const size = (x) => x.size;

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
// Tensor Transformation Functions        //
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//

export const reformat = (x, format) => format === x.dtype ? x : x.cast(format);

export const reshape = (x, shape) =>
  fn(() => x.reshape(shape.toArray())).shape_result();

export const broadcast = (x, shape) =>
  fn(() => x.broadcastTo(shape.toArray())).shape_result();

export const squeeze = (x, i) => x.squeeze(i.toArray());

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
// Tensor Logical Functions               //
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//

export const equal = (a, b) =>
  fn(() => a.equal(b).reformat_like(a)).broadcast_result();

export const not_equal = (a, b) =>
  fn(() => a.notEqual(b).reformat_like(a)).broadcast_result();

export const greater = (a, b) =>
  fn(() => a.greater(b).reformat_like(a)).broadcast_result();

export const greater_or_equal = (a, b) =>
  fn(() => a.greaterEqual(b).reformat_like(a)).broadcast_result();

export const less = (a, b) =>
  fn(() => a.less(b).reformat_like(a)).broadcast_result();

export const less_or_equal = (a, b) =>
  fn(() => a.lessEqual(b).reformat_like(a)).broadcast_result();

export const logical_and = (a, b) =>
  fn(() => a.cast("bool").logicalAnd(b.cast("bool")).reformat_like(a))
    .broadcast_result();

export const logical_or = (a, b) =>
  fn(() => a.cast("bool").logicalOr(b.cast("bool")).reformat_like(a))
    .broadcast_result();

export const logical_xor = (a, b) =>
  fn(() => a.cast("bool").logicalXor(b.cast("bool")).reformat_like(a))
    .broadcast_result();

export const logical_not = (x) => x.cast("bool").logicalNot().reformat_like(x);

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
// Tensor Arithmetic Functions            //
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//

export const add = (a, b) => fn(() => a.add(b)).broadcast_result();

export const subtract = (a, b) => fn(() => a.sub(b)).broadcast_result();

export const multiply = (a, b) => fn(() => a.mul(b)).broadcast_result();

export const divide = (a, b) =>
  fn(() => a.div(b))
    .broadcast_result()
    .map((x) => x.clip_reformat_like(a));

export const modulo = (a, b) => fn(() => a.mod(b)).broadcast_result();

export const power = (a, b) => fn(() => a.pow(b)).broadcast_result();

export const max = (a, b) => fn(() => a.maximum(b)).broadcast_result();

export const min = (a, b) => fn(() => a.minimum(b)).broadcast_result();

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
// Tensor Basic Math Functions            //
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//

export const absolute_value = tf.abs;

export const negate = tf.neg;

export const sign = tf.sign;

export const ceiling = (x) => x.cast("float32").ceil().reformat_like(x);

export const floor = (x) => x.cast("float32").floor().reformat_like(x);

export const round = tf.round;

export const exp = (x) => x.exp().clip_reformat_like(x);

export const square_root = (x) =>
  fn(() => x.cast("float32").sqrt())
    .result()
    .map((y) => y.clip_reformat_like(x));

export const ln = (x) =>
  fn(() => x.cast("float32").log())
    .result()
    .map((y) => y.clip_reformat_like(x));

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
// Tensor Reduction Functions             //
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//

export const all = (x, i) => x.cast("bool").all(i.toArray()).reformat_like(x);

export const any = (x, i) => x.cast("bool").any(i.toArray()).reformat_like(x);

export const arg_max = (x, i) => x.argMax(i).reformat_like(x);

export const arg_min = (x, i) => x.argMin(i).reformat_like(x);

export const max_over = (x, i) => x.max(i.toArray());

export const min_over = (x, i) => x.min(i.toArray());

export const sum = (x, i) => x.sum(i.toArray()).clip_reformat_like(x);

export const product = (x, i) => x.prod(i.toArray()).clip_reformat_like(x);

export const mean = (x, i) => x.mean(i.toArray()).reformat_like(x);

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
// Tensor Slicing & Joining Functions     //
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//

export const concat = (xs, i) => tf.concat(xs.toArray(), i);

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
// Tensor Conversion Functions            //
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//

export const to_float = (x) =>
  fn(() => x.clip_reformat("float32").to_number()).shape_result();

export const to_int = (x) =>
  fn(() => x.clip_reformat("int32").to_number()).shape_result();

export const to_floats = (x) => x.clip_reformat("float32").to_flat_list();

export const to_ints = (x) => x.clip_reformat("int32").to_flat_list();

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
// Tensor Utility Functions               //
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//

export function prepare_to_string(x) {
  let is_int = "int32" === x.dtype;

  let [xs, item_width] = x
    .reshape([-1])
    .arraySync()
    .reduce(
      ([xs, item_width], x) => {
        x = Number(x).toFixed(3);
        x = x.includes("e") ? x : trim_trailing(x, "0");
        if (is_int) {
          x = trim_trailing(x, ".");
        } else {
          x = x.substr(-1) === "." ? `${x}0` : x;
        }
        return [[...xs, x], Math.max(x.length, item_width)];
      },
      [[], 0],
    );

  return [toList(xs), item_width];
}

export function columns() {
  let stdout = process.stdout;
  return stdout.isTTY ? stdout.columns : 0;
}

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
// Format Functions                       //
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//

export const format_to_native = (x) => x.inspect().toLowerCase();

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
// Private Functions                      //
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//

function trim_trailing(x, character) {
  x = `${x}`;
  let i = x.length;
  while (x.charAt(--i) === character) {
    x = x.slice(0, i);
  }
  return x;
}

File: src/argamak_ffi.ex
defmodule :argamak_ffi do
  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
  # Constants                              #
  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#

  @result :gleam@result

  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
  # Tensor Creation Functions              #
  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#

  def tensor(x, format), do: fn -> Nx.tensor(x, type: format) end |> result

  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
  # Tensor Reflection Functions            #
  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#

  def size(x), do: Nx.size(x)

  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
  # Tensor Transformation Functions        #
  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#

  def reformat(x, like: y), do: Nx.as_type(x, Nx.type(y))
  def reformat(x, format), do: Nx.as_type(x, format)

  def reshape(x, shape),
    do: fn -> Nx.reshape(x, :erlang.list_to_tuple(shape)) end |> shape_result

  def broadcast(x, shape),
    do: fn -> Nx.broadcast(x, :erlang.list_to_tuple(shape)) end |> shape_result

  def squeeze(x, i), do: Nx.squeeze(x, axes: i)

  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
  # Tensor Logical Functions               #
  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#

  def equal(a, b),
    do: fn -> Nx.equal(a, b) |> reformat(like: a) end |> broadcast_result

  def not_equal(a, b),
    do: fn -> Nx.not_equal(a, b) |> reformat(like: a) end |> broadcast_result

  def greater(a, b),
    do: fn -> Nx.greater(a, b) |> reformat(like: a) end |> broadcast_result

  def greater_or_equal(a, b),
    do: fn -> Nx.greater_equal(a, b) |> reformat(like: a) end |> broadcast_result

  def less(a, b),
    do: fn -> Nx.less(a, b) |> reformat(like: a) end |> broadcast_result

  def less_or_equal(a, b),
    do: fn -> Nx.less_equal(a, b) |> reformat(like: a) end |> broadcast_result

  def logical_and(a, b),
    do: fn -> Nx.logical_and(a, b) |> reformat(like: a) end |> broadcast_result

  def logical_or(a, b),
    do: fn -> Nx.logical_or(a, b) |> reformat(like: a) end |> broadcast_result

  def logical_xor(a, b),
    do: fn -> Nx.logical_xor(a, b) |> reformat(like: a) end |> broadcast_result

  def logical_not(x), do: Nx.logical_not(x) |> reformat(like: x)

  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
  # Tensor Arithmetic Functions            #
  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#

  def add(a, b), do: fn -> Nx.add(a, b) end |> broadcast_result

  def subtract(a, b), do: fn -> Nx.subtract(a, b) end |> broadcast_result

  def multiply(a, b), do: fn -> Nx.multiply(a, b) end |> broadcast_result

  def divide(a, b),
    do: fn -> Nx.divide(a, b) end |> broadcast_result |> @result.map(&clip_reformat(&1, like: a))

  def remainder(a, b), do: fn -> Nx.remainder(a, b) end |> broadcast_result

  def power(a, b), do: fn -> Nx.pow(a, b) end |> broadcast_result

  def max(a, b), do: fn -> Nx.max(a, b) end |> broadcast_result

  def min(a, b), do: fn -> Nx.min(a, b) end |> broadcast_result

  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
  # Tensor Basic Math Functions            #
  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#

  def absolute_value(x), do: Nx.abs(x)

  def negate(x), do: Nx.negate(x)

  def sign(x), do: Nx.sign(x)

  def ceiling(x), do: Nx.ceil(x)

  def floor(x), do: Nx.floor(x)

  def round(x), do: Nx.round(x)

  def exp(x), do: Nx.exp(x) |> clip_reformat(like: x)

  def square_root(x),
    do: fn -> Nx.sqrt(x) end |> result |> @result.map(&clip_reformat(&1, like: x))

  def ln(x),
    do: fn -> Nx.log(x) end |> result |> @result.map(&clip_reformat(&1, like: x))

  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
  # Tensor Reduction Functions             #
  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#

  def all(x, i), do: Nx.all(x, axes: i) |> reformat(like: x)

  def any(x, i), do: Nx.any(x, axes: i) |> reformat(like: x)

  def arg_max(x, i), do: Nx.argmax(x, axis: i) |> reformat(like: x)

  def arg_min(x, i), do: Nx.argmin(x, axis: i) |> reformat(like: x)

  def max_over(x, i), do: Nx.reduce_max(x, axes: i)

  def min_over(x, i), do: Nx.reduce_min(x, axes: i)

  def sum(x, i), do: Nx.sum(x, axes: i) |> clip_reformat(like: x)

  def product(x, i), do: Nx.product(x, axes: i) |> clip_reformat(like: x)

  def mean(x, i), do: Nx.mean(x, axes: i) |> reformat(like: x)

  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
  # Tensor Slicing & Joining Functions     #
  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#

  def concat(xs, i), do: Nx.concatenate(xs, axis: i)

  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
  # Tensor Conversion Functions            #
  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#

  def to_float(x),
    do: fn -> clip_reformat(x, float_format(x)) |> Nx.to_number() end |> shape_result

  def to_int(x),
    do: fn -> clip_reformat(x, int_format(x)) |> Nx.to_number() end |> shape_result

  def to_floats(x), do: clip_reformat(x, float_format(x)) |> Nx.to_flat_list()

  def to_ints(x), do: clip_reformat(x, int_format(x)) |> Nx.to_flat_list()

  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
  # Tensor Utility Functions               #
  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#

  def prepare_to_string(x) do
    {format, _} = Nx.type(x)
    is_int = Enum.any?([:s, :u], fn x -> x == format end)

    x
    |> Nx.to_flat_list()
    |> Enum.reverse()
    |> Enum.reduce({[], 0}, fn x, {xs, item_width} ->
      x =
        case x do
          :infinity ->
            "Infinity"

          :neg_infinity ->
            "-Infinity"

          _else ->
            x =
              ~s(~.3#{if is_int, do: "f", else: "g"})
              |> :io_lib.format([x + 0.0])
              |> :erlang.list_to_binary()
              |> String.trim_trailing("e+0")

            x =
              if String.contains?(x, "e") do
                x
              else
                String.trim_trailing(x, "0")
              end

            if is_int do
              String.trim_trailing(x, ".")
            else
              String.replace_trailing(x, ".", ".0")
            end
        end

      {[x | xs], Kernel.max(String.length(x), item_width)}
    end)
  end

  def columns() do
    case :io.columns() do
      {:ok, columns} -> columns
      _else -> 0
    end
  end

  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
  # Format Functions                       #
  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#

  def format_to_native(x) do
    case x do
      :float64 -> {:f, 64}
      :int64 -> {:s, 64}
      :uint64 -> {:u, 64}
      :float32 -> {:f, 32}
      :int32 -> {:s, 32}
      :uint32 -> {:u, 32}
      :bfloat16 -> {:bf, 16}
      :float16 -> {:f, 16}
      :int16 -> {:s, 16}
      :uint16 -> {:u, 16}
      :int8 -> {:s, 8}
      :uint8 -> {:u, 8}
    end
  end

  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
  # Private Tensor Functions               #
  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#

  defp clip_reformat(x, like: y) do
    if Nx.type(x) == Nx.type(y) do
      # Don't clip if reformat is noop.
      x
    else
      x
      |> reformat(like: y)
      |> clip(based_on: x)
    end
  end

  defp clip_reformat(x, format) do
    x
    |> reformat(format)
    |> clip(based_on: x)
  end

  defp clip(x, based_on: y) do
    format = Nx.type(x)

    scalar = fn f ->
      format
      |> f.()
      |> Nx.from_binary(format)
      |> Nx.reshape({})
    end

    min = scalar.(&Nx.Type.min_finite_binary/1)
    max = scalar.(&Nx.Type.max_finite_binary/1)

    less = Nx.less(y, min)
    greater = Nx.greater(y, max)

    replace = fn x, predicate, with: value -> Nx.select(predicate, value, x) end

    x
    |> replace.(less, with: min)
    |> replace.(greater, with: max)
  end

  defp float_format(x), do: Nx.type(x) |> Nx.Type.to_floating()

  defp int_format(x) do
    format = Nx.type(x)
    if Nx.Type.integer?(format), do: format, else: {:s, 32}
  end

  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#
  # Private Result Functions               #
  # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#

  defp broadcast_result(f), do: result(f, or: :cannot_broadcast)

  defp shape_result(f), do: result(f, or: :incompatible_shape)

  defp result(f, opts \\ []) do
    try do
      x = f.()

      case x |> Nx.is_nan() |> Nx.any() |> Nx.to_number() do
        0 -> {:ok, x}
        1 -> {:error, :invalid_data}
      end
    rescue
      ArithmeticError -> {:error, :invalid_data}
      _else -> {:error, Keyword.get(opts, :or, :invalid_data)}
    end
  end
end
